## Assignment 1 Python Basics, Built-in Structures, Functions, Files, Numpy

This assignment will contain 5 questions with details as below. The due date is September 25 (Friday), 2020 23:59PM. Each late day will result in 20% loss of total points.

### Question 1 (20 points) Datacamp courses

Datacamp is the online learning platform for data science courses and modules. As the only collaborator in Portugal, this course provides free access to the students in Business Analytics as the complementary resource. For this assignment, you are expected to complete the following moduels and submit your completion badges:

- [Introduction to Python](https://app.datacamp.com/learn/courses/intro-to-python-for-data-science)
- [Data Types for Data Science in Python](https://app.datacamp.com/learn/courses/data-types-for-data-science-in-python)
- [Python Data Science Toolbox (Part 1)](https://app.datacamp.com/learn/courses/python-data-science-toolbox-part-1)
- [Python Data Science Toolbox (Part 2)](https://app.datacamp.com/learn/courses/python-data-science-toolbox-part-2)

Each badge will account for 5 points. If you are confident with the knowledge, you may skip the videos and directly finish the challenges.

### Question 2 (5 points) Geohashing

Geohashing is an outdoor sport activity played around the world since 2008. Contestents need to adventure to a geocoordinate generated by an algorithm that derives randomness from stock market data (as you should all know Eugene Fama's EMH). 

The algorithm is explained as below from [XKCD](https://xkcd.com/426/):
![](https://imgs.xkcd.com/comics/geohashing.png)


It described a way for a computer to create an algorithm that could generate random Global Positioning System (GPS) coordinates each day based on the Dow Jones Industrial Average and the current date.

Now your task is to decode the geohashed secret location by using Nova SBE's geocoordinate, current date (September 18, 2020) as well as the opening of DJIA on that day. Luckily, Python already has a library to calculate the geohasing using the `geohash` function in `antigravity` package:


```python
from antigravity import geohash
geohash(37.421542, -122.085589, b'2005-05-26-10458.68')
```
The first and second parameter takes the lat/lon of your preferred location and the third parameter takes the date as well as the opening of DJIA on that day, connected by `-`.

Now calculate the geohashed location using Nova SBE's geocoordinates (38.678223, -9.326266), and the current date Setember 22, 2020 and the opening of the DJIA with two decimals (you can find this at any finance news and thank you Jerome Powell!!!). 

Once you decode the location, go to google map, find out the address and write it down in the answer.

We will use the ```geohash``` function in ```antigravity``` package to calculate the geocoordinates:

In [99]:
# Question 1

from antigravity import geohash
geohash(38.678223, -9.326266, b'2020-09-22-27170.34')

38.959822 -9.819527


**Answer:** The coordinates lead to some point in the Atlantic Ocean, in front of Portugal's coast. However, when I use a thousands seperator, so "27,170.34" instead of "27170.34" as the DJIA, I get the address: Av. Gen. Humberto Delgado 54, 2840

### Question 3 (15 points) Find the truth of the universe

The number $e$, known as [Euler's number](https://en.wikipedia.org/wiki/E_(mathematical_constant)) (named after the Swiss mathematician Leonhard Euler), is a mathematical constant.  It is the base of the natural logarithm. It is the limit of $(1 + \frac{1}{n})^n$ as $n$ approaches infinity, an expression that arises in the study of compound interest, as your all know from your economics/finance 101 course.

Back in 18th century, Euler derived this number through the following equation:

$e = \frac{1}{0!}+\frac{1}{1!}+\frac{1}{2!}+\frac{1}{3!}+\frac{1}{4!}\dots+\frac{1}{n!}$

Now let's stand on the giant's shoulder to get our own $e$. To do so, you need to write a function ```factorial``` to calculate the factorial number given the $n$ and then use ```for loop``` to generate the $e$. Let $n$ go from 0 to 10, and save the derived value of $e$ in a list and print it out when n is an odd number (up to 10).

Hints: 
Consider use range(#begin, #end) method

In [100]:
# Question 2

# defining the function "factorial", to calculate the factorial number of an input number n
def factorial(n):
    start = 1
    for i in range(1,n+1):
        start *= i
    return start

# creating an empty list and filling it with all Euler numbers of odd n's in a range from 0-10
Euler_list = []
sum = 0
for n in range(11):
    e = 1 / (factorial(n))
    sum += e
    if n % 2 != 0:
        Euler_list.append(sum)
print(Euler_list)

[2.0, 2.6666666666666665, 2.7166666666666663, 2.7182539682539684, 2.7182815255731922]


### Question 4 (20 points) The monologue of Macbeth

The file ```macbeth.txt``` contains the act of Macbeth, one of the greatest pieces of William Shakespeare. Read the file as we show in the lecture and answer the following question. 

1. Print the number of lines in the file
2. Identify how many times that Macebeth narrated, that is the number of times the text shows ```MACBETH``` (in upper case), not the number of times ```Macbeth```  appears in the text.
3. Slice the file with content only from the monologue of Macbeth and save it to a new file ```macbeth_monologue.txt```. 

Hints: you can check each line of the text and start to save the text to a new list when the line contains `MACBETH` and stop to save when the line contains another line of character name.

**Note** If you are using Google Colab, you need first to upload this file to the computing environment using the following code:
```python
from google.colab import files

files.upload()
```

Then it will ask you to upload the file, select the file and it will be uploaded. You can verify whether it has been successfully uploaded at the left of the notebook `Files`

In [101]:
# Question 3.1

# use the len() function on readlines() to get the number of lines
path = "macbeth.txt"
with open(path) as file:
    print("Number of lines: " + str(len(file.readlines())))

Number of lines: 3375


In [102]:
# Question 3.2

# use the count() function to detect all lines where "MACBETH\n" appears
with open(path) as file:
    read_data = file.readlines()
    print("Number of Macbeth narrations: " + str(read_data.count("MACBETH\n")))

Number of Macbeth narrations: 145


In [103]:
# Question 3.3

# create a list, only storing the indexes of the lines where "MACBETH\n" appears
mon_lines = []
with open(path, 'r') as myFile:
    for num, line in enumerate(myFile, 1):
        if (line == "MACBETH\n"):
            mon_lines.append(num)
            
# create a list, storing all lines between the lines where "MACBETH\n" appears and the lines where only uppercases appear
monolouge = []
j = 0
with open(path, 'r') as full:   
    lines = full.readlines()
    for j in mon_lines:
        i = j
        mb = True
        while mb:
            if (lines[i].isupper() == False):
                monolouge.append(lines[i])
                i += 1
            else:
                mb = False

# save this list into a new text file
with open('macbeth_monologue.txt', 'w') as mb_file:
    mb_file.writelines(monolouge)

### Question 5 (40 points)  Is it Instagrammable?

Consider any photo you take at Carcavelos beach:

![carcavelos](carcavelos.jpg)

<br>

An image is composed of three matrixes, each one for an RGB (red, green and blue) channel. Each matrix has values between 0 and 255.


![image](https://static.packt-cdn.com/products/9781789613964/graphics/e91171a3-f7ea-411e-a3e1-6d3892b8e1e5.png)

On this exercise, you will apply a filter to the image in a process called convolution. This process uses the filter as a matrix that is multiplied as a sliding window to an image. On the following animation, the filter is represented as yellow, the image channel as green, and the convolution result as red.

![convolution](https://icecreamlabs.com/wp-content/uploads/2018/08/33-con.gif)


Note that the resulted matrix has a smaller shape than the original, to ensure the same shape consider a padding with value 0 that enables the filter to slide throw every pixel value as shown in the following image:

<br>

![padding](https://media5.datahacker.rs/2018/11/sl_1.png)

<br><br>

This question requires the `numpy` and `Pillow` libraries that may already be installed on your enviroment. However, if you have a problem importing this libraries, install it via pip. Example for Pillow: `pip install Pillow` 

<br>

Consider the following code to convert an image into a numpy array:

```python
from PIL import Image
from numpy import asarray

# load the image
image = Image.open('carcavelos.jpg')

# convert image to numpy array
data = asarray(image)
```



You can save a numpy array as an image with the following code:

```python
"""
color image
"""

# convert numpy array to image
img = Image.fromarray(data, 'RGB')

# save image and .png
img.save('image.png')

"""
grayscale image
"""

# convert numpy array to image
data = data.astype(np.uint8)
img = Image.fromarray(data)

# save grayscale image as .png
img.save('image.png')
```

In [104]:
from PIL import Image
from numpy import asarray

**Question 5.1** (5 points) What is the shape of the image? Answer with the shape and indicate what each dimension represents.

In [105]:
# load the image
image = Image.open('carcavelos.jpg')

# convert image to numpy array
data = asarray(image)

# print the shape of the image
print("Shape of the image: " + str(data.shape))

Shape of the image: (816, 1226, 3)


**Answer:** 816 represents the height of the image (in pixels), 1226 represents the width of the image (in pixels) and the 3 represents the image's 3 colour channels RGB (red, green, blue).

**Question 5.2** (15 points) Save the "carcavelos.jpg" image as a grayscale image into "carcavelos-grayscale.png". The image should result from a numpy array with only one channel instead of 3, consider the following weights for each channel: 0.30*R + 0.59*G + 0.11*B

In [106]:
from matplotlib.image import imread

# read coloured image and define 3 variables R,G and B, referring to the 3 colour channels of the image
image = imread("carcavelos.jpg")
R,G,B = image[:,:,0], image[:,:,1], image[:,:,2]

# define the 3 variables R_weight, G_weight and B_weight to weigh each channel for the conversion
R_weight, G_weight, B_weight = 0.3, 0.59, 0.11

# create the new array grayscale, multiplying each numpy array R, G and B with their corresponding weights and adding them together
grayscale = R_weight*R + G_weight*G + B_weight*B

# convert numpy array to image
grayscale = grayscale.astype(np.uint8)
img = Image.fromarray(grayscale)

# save image as .png
img.save("carcavelos-grayscale.png")

**Question 5.3** (25 points) Apply convolution operations as described above to "carcavelos.jpg". 
1. Save an image into "carcavelos-sharpen.png" with the applied filter: `np.array([[0,-1,0],[-1,5,-1],[0,-1,0]])`
2. Save an image into "carcavelos-blur.png" with the applied filter: `np.array([[1,1,1],[1,1,1],[1,1,1]])/9`

In [107]:
import scipy
from scipy import signal
import numpy as np

# load the image
image = Image.open('carcavelos.jpg')

# convert image to numpy array
data = asarray(image)

# 1. sharpening:

# create the variable sharpen, storing the filter array 
sharpen = np.array([[0,-1,0],[-1,5,-1],[0,-1,0]])

# reshape sharpen from a 2-dimensional 3x3 array to a 3-dimensional 3x3x1 array
sharpen = sharpen.reshape((sharpen.shape[0], sharpen.shape[1], 1))

# convolve the original image by using the sharpening filter
conv_sharp = scipy.signal.convolve(data, sharpen)

# remove RGB-values higher than 255 and lower than 0 to prevent oversharpening the image
conv_sharp = np.where(conv_sharp > 255, 255, conv_sharp)
conv_sharp = np.where(conv_sharp < 0, 0, conv_sharp)

# convert numpy array to image
conv_sharp = conv_sharp.astype(np.uint8)
img_sharp = Image.fromarray(conv_sharp)

# save image as .png
img_sharp.save("carcavelos-sharpen.png")


# 2. blurring:

# create the variable blur, storing the filter array 
blur = np.array([[1,1,1],[1,1,1],[1,1,1]])/9

# reshape blur from a 2-dimensional 3x3 array to a 3-dimensional 3x3x1 array
blur = blur.reshape((blur.shape[0], blur.shape[1], 1))

# convolve the original image by using the blurring filter
conv_blur = scipy.signal.convolve(data, blur)

# convert numpy array to image
conv_blur = conv_blur.astype(np.uint8)
img_blur = Image.fromarray(conv_blur)

# save image as .png
img_blur.save("carcavelos-blur.png")